### PR TITLE
chore(component-testing): Converting @cypress/react to use dev server plugin

### PR DESCRIPTION
**Pull request by JessicaSachs**
After merging, you can run `cy:open` and `cy:run` from the `npm/react` package.

Beware, in Open Mode the socket disconnect bug (currently CT-189) requires you to wait until the bundle is finished before interacting with the UI. Run mode does not have this problem. This bug is not as visible in `npm/vue` because the bundle is significantly smaller (2s instead of ~22s for React)

* adds cy:run and cy:open tasks to react
* hooks up `@cypress/react`'s component tests to use the dev server plugin.
* turns off a lot of the `specType === 'component'` checks
* moved hook logic from support.js to the top of the mount utility
* fixes existing cy:run task in Vue
* renaming `devserver` to `devServer (variables)` or `dev-server (strings, filenames)` for consistency

### CODE REVIEW

File: npm/webpack-dev-server/src/startServer.ts [hunk 1]
```diff
@@ -2,17 +2,23 @@ import webpack from 'webpack'
...
-  const foundConfig = fw.getWebpackOptions()
-  const webpackConfig = await makeWebpackConfig(foundConfig || userWebpackConfig, {
+export async function start (initialWebpackConfig, { specs, config, devServerEvents }): Promise<WebpackDevServer> {
+  if (initialWebpackConfig == null) {
+    debug('User did not pass in any webpack configuration')
+    // TODO: implement a method in find-webpack to return the path where it found the configuration
+    initialWebpackConfig = fw.getWebpackOptions()
+  }
+
+  const webpackConfig = await makeWebpackConfig(initialWebpackConfig, {
```
Comment (JessicaSachs) [inline npm/webpack-dev-server/src/startServer.ts:17]: Bug fix, we stopped preferring the user's configuration and instead always used findWebpack 😬

File: npm/webpack-dev-server/src/startServer.ts [hunk 2]
```diff
@@ -21,5 +27,5 @@ export async function start (userWebpackConfig = {}, { specs, config, devserverE
 
   debug('starting webpack dev server')
 
-  return new WebpackDevServer(compiler, { hot: true, noInfo: !debug.enabled })
+  return new WebpackDevServer(compiler, { hot: true })
```
Comment (JessicaSachs) [inline npm/webpack-dev-server/src/startServer.ts:30]: Now that we have to wait for the dev server to finish compiling (due to the race condition bug), seeing the dev server is very important

File: npm/react/cypress/support/index.js [hunk 3]
```diff
@@ -15,5 +15,3 @@
 
 // custom commands provided by this package, built from TypeScript code in "lib"
 // using "npm run transpile"
-import '@cypress/react/hooks'
-import 'cypress-react-selector'
```
Comment (JessicaSachs) [inline npm/react/cypress/support/index.js:19]: We don't import support right now. (honestly, we shouldn't be running this hook for _every_ test anyways... only the ones that use mount)

I moved these into the top of `mount.ts` which seems more appropriate

File: npm/react/src/hooks.ts [hunk 4]
```diff
@@ -1,7 +1,7 @@
 // @ts-ignore
 const unfetch = require('unfetch/dist/unfetch.js')
 // @ts-ignore
-const isComponentSpec = () => Cypress.spec.specType === 'component'
+const isComponentSpec = () => true // TODO: refactor code depending on this fn
```
Comment (dmtrKovalenko) [inline npm/react/src/hooks.ts:4]: Not sure what you mean by this comment

File: npm/react/package.json [hunk 5]
```diff
@@ -6,7 +6,10 @@
   "scripts": {
     "build": "rimraf dist && yarn transpile",
     "build-prod": "yarn build",
-    "cy:open": "node ../../scripts/cypress open",
+    "cy:open": "node ../../scripts/start.js --component-testing --project ${PWD}",
+    "cy:open:debug": "NODE_OPTIONS=--max-http-header-size=1048576 node --inspect-brk ../../scripts/start.js --component-testing --project ${PWD}",
+    "cy:run": "node ../../scripts/start.js --component-testing --run-project ${PWD}",
+    "cy:run:debug": "NODE_OPTIONS=--max-http-header-size=1048576 node --inspect-brk ../../scripts/start.js --component-testing --run-project ${PWD}",
```
Comment (dmtrKovalenko) [inline npm/react/package.json:12]: Should we use vscode terminal tasks instead of npm scripts for this?

File: npm/react/cypress/plugins/index.js [hunk 6]
```diff
@@ -55,15 +54,11 @@ const webpackOptions = {
...
-const options = {
-  // send in the options from your webpack.config.js, so it works the same
-  // as your app's code
-  webpackOptions,
-  watchOptions: {},
-}
-
-module.exports = (on, config) => {
-  on('file:preprocessor', webpackPreprocessor(options))
+const cypressPluginsFn = (on, config) => {
```
Comment (dmtrKovalenko) [inline npm/react/cypress/plugins/index.js:57]: For me this looks a bit confusing in comparison with other cypress examples. Maybe we can do it simply:

```suggestion
module.exports = (on, config) => {
```

**Review by dmtrKovalenko — COMMENTED** [review]
We still need to update `plugins` folder for different frameworks like react-scirpts, nextjs and others. Should we make this in different PR?

**Comment by dmtrKovalenko** [comment]
Related to the hooks, I don't think we need them at all – these extra polyfills and renderers/cleaners should be as a part of the component testing `runner` frontend.  They don't make a lot of sense when we are making component testing as special mode inside cypress

File: npm/react/package.json (hunk 5 above)
Comment (lmiller1990) [reply npm/react/package.json:12]: Should move these to `packages/server-ct` and have something like `cy:open:react`, `cy:open:vue` etc.

File: npm/react/src/hooks.ts (hunk 4 above)
Comment (JessicaSachs) [reply npm/react/src/hooks.ts:4]: specType doesn't exist any more -- this was when we were mixing the mount commands with integration specs. We need to go through and fix specType not being sent through to mobx.

File: npm/react/package.json (hunk 5 above)
Comment (JessicaSachs) [reply npm/react/package.json:12]: Yeah I think we should hook up all of these in server-ct shortly.

**Comment by JessicaSachs** [comment]
> We still need to update `plugins` folder for different frameworks like react-scirpts, nextjs and others. Should we make this in different PR?

Yeah. All of the plugins changes need to happen still.

> Related to the hooks, I don't think we need them at all – these extra polyfills and renderers/cleaners should be as a part of the component testing runner frontend. They don't make a lot of sense when we are making component testing as special mode inside cypress

Yep, agreed.

**Comment by JessicaSachs** [comment]
We're going to keep this open until lachlan has a chance to deal with all the merge conflicts in the e2e branch and the current evergreen-bem branch

**Comment by lmiller1990** [comment]
@JessicaSachs merging e2e may not happen - https://github.com/cypress-io/cypress/pull/14266

This code looks mostly fine to me, I have to test it locally. I think we need to start blocking merging on any PR without the CI been green. I am not sure what it would take to get CI passing here.

**Pull request merged**